src/metadata-extractors.ts

```diff
@@ -17,88 +19,34 @@ export interface ClassInfo {
-): ClassInfo | undefined {
+export function getClassInfo(classNode: ts.ClassDeclaration & ts.Node, checker: ts.TypeChecker): ClassInfo | undefined {
   if (!classNode.name) return
 
-  const properties = checker
-    .getPropertiesOfType(checker.getTypeAtLocation(classNode))
-    .map((prop) => {
-      return getPropertyInfo(prop, context, checker)
-    })
+  const node = createWrappedNode<ts.Node>(classNode, { typeChecker: checker }).asKindOrThrow(
+    SyntaxKind.ClassDeclaration
+  )
 
   return {
-    name: classNode.name.getText(),
-    fields: properties.filter((prop) => !prop.isMethod),
-    methods: properties.filter((prop) => prop.isMethod),
+    name: node.getNameOrThrow(),
+    fields: node.getInstanceProperties().map((p) => ({ name: p.getName(), typeInfo: getTypeInfo(p.getType()) })),
+    methods: node.getInstanceMethods().map((m) => ({ name: m.getName(), typeInfo: getTypeInfo(m.getReturnType()) })),
   }
 }
 
-function getPropertyInfo(
-  prop: ts.Symbol,
-  context: ts.TransformationContext,
-  checker: ts.TypeChecker
-): PropertyInfo {
-  return {
-    name: prop.getName(),
-    isMethod: ts.isMethodDeclaration(prop.valueDeclaration),
-    typeInfo: getTypeInfo(prop.valueDeclaration, context),
-  }
-}
-
-function getTypeInfo(
-  node: ts.Node,
-  context: ts.TransformationContext
-): TypeInfo {
+function getTypeInfo(type: Type): TypeInfo {
+  const isNullable = type.isNullable()
+  type = type.getNonNullableType()
   const typeInfo: TypeInfo = {
-    name: 'undefined',
-    parameters: [],
+    name: type.getSymbol()?.getName() || type.getText(), // getSymbol() is used for complex types, in which cases getText() returns too much information (e.g. Map<User> instead of just Map)
+    isNullable: isNullable,
+    isEnum: type.isEnum(),
+    parameters: type.getTypeArguments().map((a) => getTypeInfo(a)),
   }
-  if (hasNoTypeInfo(node)) {
-    return typeInfo
-  }
-
-  function visitor(node: ts.Node): ts.VisitResult<ts.Node> {
-    if (!ts.isTypeNode(node)) {
-      return ts.visitEachChild(node, visitor, context)
-    }
-    if (ts.isTypeReferenceNode(node)) {
-      typeInfo.name = node.typeName.getText()
-      typeInfo.parameters =
-        node.typeArguments?.map((node) => getTypeInfo(node, context)) ?? []
-    } else if (ts.isFunctionTypeNode(node)) {
-      typeInfo.name = 'Function' // TODO: We could get more detailed here
-    } else if (ts.isArrayTypeNode(node)) {
-      typeInfo.name = Array.name
-      typeInfo.parameters = [getTypeInfo(node.elementType, context)]
-    } else {
-      typeInfo.name = normalizeTypeName(node.getText())
-    }
-
-    return node
+  if (typeInfo.isEnum) {
+    typeInfo.parameters = type
+      .getUnionTypes()
+      .map((t) => ({ name: t.getSymbol()?.getName() || t.getText(), isNullable: false, isEnum: false, parameters: [] }))
   }
-
-  ts.visitNode(node, visitor)
+  if (typeInfo.name.length > 0) typeInfo.name = typeInfo.name[0].toUpperCase() + typeInfo.name.slice(1)
```

Comment:
I guess this is, especially, for primitive types like `boolean`, `number`, etc. Correct?

Another question: Do you know if `typesInfo.name.length` can be empty or null?